File: control_plane/runtime/contracts/challenges.py
"""Public HTTP contracts for the challenge catalogue."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class ResourceBudget(BaseModel):
    cpuMillis: int = Field(ge=1)
    memoryMiB: int = Field(ge=1)


class OriginRef(BaseModel):
    provider: str | None = None
    link: str | None = None


class ChallengeCard(BaseModel):
    key: str
    shortCode: str
    name: str
    level: str
    score: int | None = None
    acceptedCount: int = 0
    labels: list[str] = Field(default_factory=list)
    mode: str = "stdio"
    runtimes: list[str] = Field(default_factory=list)
    budget: ResourceBudget
    customChecker: bool = False
    interactive: bool = False
    origin: OriginRef | None = None


class ExampleCase(BaseModel):
    stdin: str = ""
    stdout: str = ""
    explanation: str | None = None


class ChallengePrompt(BaseModel):
    overview: str = ""
    inputContract: str | None = None
    outputContract: str | None = None
    interactionContract: str | None = None
    notes: str | None = None
    markdown: str | None = None


class EvaluationPolicy(BaseModel):
    inputMode: str = "stdio"
    inputFile: str | None = None
    outputFile: str | None = None
    checker: str = "tokens"
    checkerOptions: dict[str, Any] | None = None
    testCount: int | None = None
    visibleTests: int | None = None


class SourcePolicy(BaseModel):
    artifactType: str = "source"
    extensions: list[str] = Field(default_factory=list)
    entrypoint: str | None = None


class ChallengeView(ChallengeCard):
    derivedBudget: bool = False
    prompt: ChallengePrompt = Field(default_factory=ChallengePrompt)
    examples: list[ExampleCase] = Field(default_factory=list)
    evaluation: EvaluationPolicy = Field(default_factory=EvaluationPolicy)
    sourcePolicy: SourcePolicy = Field(default_factory=SourcePolicy)
    referenceSolutions: int = 0


class PageInfo(BaseModel):
    index: int
    size: int
    totalEntries: int
    totalPages: int


class ChallengeCollection(BaseModel):
    entries: list[ChallengeCard]
    pageInfo: PageInfo


ChallengeOrder = Literal["shortCode", "name", "level", "acceptedCount", "key", "score", "mode"]

File: web_console/lib/contracts/challenges.ts
export type ChallengeLevel = "easy" | "medium" | "hard";
export type ChallengeMode = "stdio" | "interactive" | "library" | "sql" | "bash" | "archive";

export interface ResourceBudget {
  cpuMillis: number;
  memoryMiB: number;
}

export interface OriginRef {
  provider?: string | null;
  link?: string | null;
}

export interface Challenge {
  key: string;
  shortCode: string;
  name: string;
  level: ChallengeLevel;
  score?: number | null;
  acceptedCount: number;
  labels: string[];
  mode: ChallengeMode;
  runtimes: string[];
  budget: ResourceBudget;
  customChecker: boolean;
  interactive: boolean;
  origin?: OriginRef | null;
}

export interface ExampleCase {
  stdin: string;
  stdout: string;
  explanation?: string | null;
}

export interface ChallengePrompt {
  overview: string;
  inputContract?: string | null;
  outputContract?: string | null;
  interactionContract?: string | null;
  notes?: string | null;
  markdown?: string | null;
}

export interface EvaluationPolicy {
  inputMode: string;
  inputFile?: string | null;
  outputFile?: string | null;
  checker: string;
  checkerOptions?: Record<string, unknown> | null;
  testCount?: number | null;
  visibleTests?: number | null;
}

export interface SourcePolicy {
  artifactType: string;
  extensions: string[];
  entrypoint?: string | null;
}

export interface ChallengeView extends Challenge {
  derivedBudget: boolean;
  prompt: ChallengePrompt;
  examples: ExampleCase[];
  evaluation: EvaluationPolicy;
  sourcePolicy: SourcePolicy;
  referenceSolutions: number;
}

export interface PageInfo {
  index: number;
  size: number;
  totalEntries: number;
  totalPages: number;
}

export interface ChallengeCollection {
  entries: Challenge[];
  pageInfo: PageInfo;
}

export type ChallengeOrder = "shortCode" | "name" | "level" | "acceptedCount" | "key" | "score" | "mode";

export interface ChallengeFilter {
  search?: string;
  level?: ChallengeLevel | "all";
  label?: string | "all";
  mode?: ChallengeMode | "all";
  orderBy?: ChallengeOrder;
  direction?: "asc" | "desc";
  page?: number;
  size?: number;
}

const DEFAULT_BUDGET: ResourceBudget = { cpuMillis: 2000, memoryMiB: 256 };

export function challengeBudget(challenge: Pick<Challenge, "budget">): ResourceBudget {
  return challenge.budget ?? DEFAULT_BUDGET;
}
